fix(ivy): component destroy hook called twice when configured as provider

crisbeto opened the pull request:
Fixes the `ngOnDestroy` hook on a component or directive being called twice, if the type is also registered as a provider.

This PR resolves FW-1010.

crisbeto committed:
fix(ivy): component destroy hook called twice when configured as provider

Fixes the `ngOnDestroy` hook on a component or directive being called twice, if the type is also registered as a provider.

This PR resolves FW-1010.

kara commented on packages/core/test/render3/component_spec.ts:
```diff
@@ -703,4 +703,36 @@ describe('view destruction', () => {
+        type: ComponentWithOnDestroy,
+        consts: 0,
+        vars: 0,
+        factory: () => new ComponentWithOnDestroy(),
```
The `directiveInject` in the template below should be moved here (this is what the compiler actually generates)

```suggestion
        factory: () => new ComponentWithOnDestroy(directiveInject(testToken),
```

crisbeto replied:
It can't be moved there, because it generates a circular dependency (`ComponentWithOnDestroy` is the one being provided as the `testToken`).

kara commented on packages/core/src/render3/di.ts:
```diff
@@ -514,7 +514,10 @@ export function getNodeInjectable(
-        (tView.destroyHooks || (tView.destroyHooks = [])).push(index, value.ngOnDestroy);
+        const type = value.constructor as Type<any>;
+        if (!getComponentDef(type) && !getDirectiveDef(type)) {
+          (tView.destroyHooks || (tView.destroyHooks = [])).push(index, value.ngOnDestroy);
```
I know this was existing logic, but I just noticed that we are pushing to `tView.destroyHooks` on every pass, not just the first one. We should wrap that logic in a guard to prevent duplicate pushes:

```suggestion
if (tView.firstTemplatePass) {
   (tView.destroyHooks || (tView.destroyHooks = [])).push(index, value.ngOnDestroy);
}
```

kara replied:
Update: looks like it will be a more complicated fix because of things like `injector.get`, so we are punting as out of scope for this PR

crisbeto commented:
The feedback has been addressed.

crisbeto committed:
fixup! fix(ivy): component destroy hook called twice when configured as provider

crisbeto commented:
@kara I've reworked it based on the design. Can you take another look?

kara commented on packages/core/test/render3/providers_spec.ts:
```diff
@@ -1355,6 +1355,218 @@ describe('providers', () => {
       expect(logs).toEqual(['Injectable OnDestroy']);
     });
+
+    it('should not inherit ngOnDestroy hooks on providers', () => {
```
We are trying to avoid adding any new tests with generated code written by hand. Can you convert these into TestBed tests and move them to the `core/test/acceptance` folder?

kara commented on packages/core/src/render3/di_setup.ts:
```diff
@@ -81,10 +81,18 @@ function resolveProvider(
 
+    if (isClassProvider(provider) || isTypeProvider(provider)) {
+      const prototype = ((provider as ClassProvider).useClass || provider).prototype;
+      if (prototype.hasOwnProperty('ngOnDestroy')) {
```
Nit: Maybe a quick comment here to explain that providers should not inherit `ngOnDestroy` hooks from superclasses, hence the `hasOwnProperty` check?

kara reviewed (changes requested):
Looks good. Just needs the tests to be converted to TestBed.

kara commented:
@crisbeto There should also be a number of Material tests fixed by this PR. Could you remove them from the material blocklist here:

https://github.com/angular/angular/blob/master/tools/material-ci/angular_material_test_blocklist.js

filipesilva reviewed (approved):
The changes in `tools/material-ci/angular_material_test_blocklist.js` LGTM.

crisbeto commented:
@kara I've moved the unit tests to `TestBed` and I've updated the blocklist, however switching to `TestBed` revealed a discrepancy between Ivy and ViewEngine in these two tests: 

* `should not inherit ngOnDestroy hooks on providers`
* `In Ivy the destroy hook of useClass provider is invoked`

I've marked them as "only in Ivy", but I think that it's up for discussion how we should handle them.

kara commented on packages/core/test/acceptance/providers_spec.ts:
```diff
@@ -0,0 +1,161 @@
+      expect(logs).toEqual(['OnDestroy']);
+    });
+
+    onlyInIvy('In Ivy the destroy hook of useClass provider is invoked')
```
Yeah, this is a [View Engine bug](https://github.com/angular/angular/issues/14821) that manifests when the token and value differ.  It should work in both when the token/value are the same. Maybe add another test for that?

kara commented on packages/core/test/acceptance/providers_spec.ts:
```diff
@@ -0,0 +1,161 @@
+
+describe('providers', () => {
+  describe('lifecycles', () => {
+    onlyInIvy('In Ivy providers do not inherit ngOnDestroy hooks')
```
This test was based on my initial testing of provider inheritance, but I think I confused this with something else. We should make this work how it does in View Engine, so let's flip this and remove the `hasOwnProperty` check.

crisbeto commented:
I've updated it based on our discussion @kara.

crisbeto committed:
fixup! fix(ivy): component destroy hook called twice when configured as provider

kara reviewed (approved):
LGTM

kara commented:
[presubmit](http://test/OCL:234990533:BASE:234990572:1550762577962:4c6d2adb)

kara commented:
merge assistance: global approval
